### PR TITLE
Feature/35929 pupil group database changes

### PR DESCRIPTION
**Pull request by GuyHarwood**
Shifts from many groups for many pupils to one group per pupil, which matches specification.

- [x] introduce nullable group_id reference column in pupil table
- [x] manual sql migration to migrate existing groups (for reference and potential later use)
- [x] modify data access components and services to use new design
- [x] update dependent views and procedures
- [x] migrate existing group assignments to new model
- [x] remove soft deletes [optional]

### CODE REVIEW

File: admin/data/sql/migrations/20190815150258.do.pupilgroup-vewpupilseligibleforlivepingeneration.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
+ALTER VIEW [mtc_admin].vewPupilsEligibleForLivePinGeneration AS
+  SELECT
+         p.id,
+         p.foreName,
+         p.middleNames,
+         p.lastName,
+         p.dateOfBirth,
+         p.urlSlug,
+         p.school_id,
+         p.group_id,
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190815150258.do.pupilgroup-vewpupilseligibleforlivepingeneration.sql]: @jon-shipley soft deletes now removed

File: admin/data/sql/migrations/20190815114552.do.pupil-group-column.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+alter table mtc_admin.pupil
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190815114552.do.pupil-group-column.sql]: idempotent migrations will be a requirement after baselining in september

File: admin/data/sql/migrations/20190816091654.do.create-pupilgroup-archive.sql [hunk 3]
```diff
@@ -0,0 +1,9 @@
+
+create table mtc_admin.z_pupilGroup_archive
```
Comment (GeorgeChatzigiannis) [inline admin/data/sql/migrations/20190816091654.do.create-pupilgroup-archive.sql]: @GuyHarwood why not `SELECT INTO` z_pupilGroup_archive from pupilGroup? That would save 1 migration

File: admin/data/sql/migrations/20190815150016.do.pupilgroup-vewpupilregister.sql [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+         ISNULL(g.name, '-') as groupName,
+         ps.code as pupilStatusCode,
+         lastCheck.id as lastCheckId,
+         cs.code as lastCheckStatusCode,
+         lastPupilRestart.id as pupilRestartId,
+         lastPupilRestart.check_id as pupilRestartCheckId
+
+  FROM
+       [mtc_admin].[pupil] p LEFT JOIN
+       [mtc_admin].[group] g ON (p.group_id = g.id) JOIN
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20190815150016.do.pupilgroup-vewpupilregister.sql]: Why is this join to [group] needed?

File: admin/data/sql/migrations/20190815150258.do.pupilgroup-vewpupilseligibleforlivepingeneration.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (jon-shipley) [reply admin/data/sql/migrations/20190815150258.do.pupilgroup-vewpupilseligibleforlivepingeneration.sql]: 🏆

File: admin/data/sql/migrations/20190816105838.do.create-group-archive.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+create table mtc_admin.[z_group_archive]
```
Comment (GeorgeChatzigiannis) [inline admin/data/sql/migrations/20190816105838.do.create-group-archive.sql]: SELECT INTO  can be used for this one too.

File: admin/data/sql/migrations/20190815151201.undo.pupilgroup-vewpupilswithactivefamiliarisationpins.sql [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+         p.middleNames,
+         p.dateOfBirth,
+         p.urlSlug,
+         p.school_id,
+         pin.val as pin,
+         cp.pinExpiresAt,
+         g.group_id
+  FROM [mtc_admin].[pupil] p
+         INNER JOIN [mtc_admin].[school] s ON (p.school_id = s.id)
+         LEFT JOIN  [mtc_admin].[pupilGroup] g ON (g.pupil_id = p.id)
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20190815151201.undo.pupilgroup-vewpupilswithactivefamiliarisationpins.sql]: Why do we need to join to [group] here?

File: admin/data/sql/migrations/20190815151308.undo.pupilgroup-vewpupilswithactivelivepins.sql [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+         p.middleNames,
+         p.dateOfBirth,
+         p.urlSlug,
+         p.school_id,
+         pin.val as pin,
+         cp.pinExpiresAt,
+         g.group_id
+  FROM [mtc_admin].[pupil] p
+         INNER JOIN [mtc_admin].[school] s ON p.school_id = s.id
+         LEFT JOIN  [mtc_admin].[pupilGroup] g ON g.pupil_id = p.id
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20190815151308.undo.pupilgroup-vewpupilswithactivelivepins.sql]: Why do we need to Join to [group] here?

File: admin/data/sql/migrations/20190816091709.undo.migrate-pupilgroup-to-archive.sql [hunk 8]
```diff
@@ -0,0 +1,4 @@
+SET IDENTITY_INSERT mtc_admin.pupilGroup ON
```
Comment (jon-shipley) [inline admin/data/sql/migrations/20190816091709.undo.migrate-pupilgroup-to-archive.sql]: 💯

File: admin/services/group.service.js [hunk 9]
```diff
@@ -41,14 +41,14 @@ groupService.getGroupsAsArray = async function (schoolId) {
 /**
  * Get pupils filtered by schoolId and groupId.
  * @param schoolId required.  the school context
- * @param groupIdToExclude optionally exclude a single group from the returned set
+ * @param groupIdToInclude optionally exclude a single group from the returned set
```
Comment (jon-shipley) [inline admin/services/group.service.js]: Param name does not match description: include / exclude

File: admin/data/sql/migrations/20190815144902.do.pupilgroup-spgetpupilsresults.sql [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        lastPupilRestart.id as pupilRestartId,
+        lastPupilRestart.check_id as pupilRestartCheckId,
+        latestPupilCheck.mark,
+        latestPupilCheck.maxMark,
+        cs.code as checkStatusCode,
+        ac.reason
+    FROM [mtc_admin].[pupil] p
+    JOIN [mtc_admin].[pupilStatus] ps
+        ON (p.pupilStatus_id = ps.id)
+    LEFT JOIN [mtc_admin].[group] g
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190815144902.do.pupilgroup-spgetpupilsresults.sql]: good spot 🏆

File: admin/data/sql/migrations/20190816091654.do.create-pupilgroup-archive.sql (hunk 3 above)
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190816091654.do.create-pupilgroup-archive.sql]: that is used in the next migration to do the data copy.  i like to keep simple units of work per migration.  just my preference 🤷‍♂

File: admin/data/sql/migrations/20190816105838.do.create-group-archive.sql (hunk 5 above)
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190816105838.do.create-group-archive.sql]: as above.

File: admin/data/sql/migrations/20190815144902.do.pupilgroup-spgetpupilsresults.sql (hunk 10 above)
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190815144902.do.pupilgroup-spgetpupilsresults.sql]: good spot 🏆

File: admin/data/sql/migrations/20190815151201.undo.pupilgroup-vewpupilswithactivefamiliarisationpins.sql (hunk 6 above)
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190815151201.undo.pupilgroup-vewpupilswithactivefamiliarisationpins.sql]: thats the undo query, which does in fact need the group table.

File: admin/data/sql/migrations/20190815151308.undo.pupilgroup-vewpupilswithactivelivepins.sql (hunk 7 above)
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190815151308.undo.pupilgroup-vewpupilswithactivelivepins.sql]: thats the undo query, which does in fact need the group table.

File: admin/services/group.service.js (hunk 9 above)
Comment (GuyHarwood) [reply admin/services/group.service.js]: fixed

File: admin/data/sql/migrations/20190815150016.do.pupilgroup-vewpupilregister.sql (hunk 4 above)
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190815150016.do.pupilgroup-vewpupilregister.sql]: `g.name` is used in the select

File: admin/data/sql/migrations/20190816091654.do.create-pupilgroup-archive.sql (hunk 3 above)
Comment (GuyHarwood) [reply admin/data/sql/migrations/20190816091654.do.create-pupilgroup-archive.sql]: great idea 🏆

**Comment by GuyHarwood** [comment]
There is currently an access failure on `dot.net` when building containers in travis.  i have reproduced this locally, so it should just be a transient issue which will hopefully resolve soon.

**Pull request merged**